pkg/api/index.go

```diff
@@ -383,6 +383,35 @@ func (hs *HTTPServer) getNavTree(c *models.ReqContext, hasEditPerm bool) ([]*dto
 		Children:   []*dtos.NavLink{},
 	})
 
+	if hs.Features.IsEnabled(featuremgmt.FlagNewNavigation) {
+		return hs.setNavPreferences(c, navTree)
+	}
+
+	return navTree, nil
+}
+
+func (hs *HTTPServer) setNavPreferences(c *models.ReqContext, navTree []*dtos.NavLink) ([]*dtos.NavLink, error) {
+	// query preferences table for any navbar preferences
+	prefsQuery := models.GetPreferencesJsonDataWithDefaultsQuery{User: c.SignedInUser}
+	if err := hs.SQLStore.GetPreferencesJsonDataWithDefaults(c.Req.Context(), &prefsQuery); err != nil {
+		return nil, err
+	}
+	jsonDataPref := prefsQuery.Result.JsonData
+
+	if jsonDataPref != nil {
+		for _, navItem := range navTree {
+			// Set any that exist in the navbar preferences to hide=false
+			// Fix me: how can we type this json properly
+			for _, pref := range jsonDataPref.Get("navbar").MustArray() {
+				typedPref := pref.(map[string]interface{})
+				if navItem.Id == typedPref["id"] {
+					navItem.HideFromNavbar = typedPref["hideFromNavbar"].(bool)
+					break
+				}
+			}
```

Comment:
nooby go/json question: how do we type this correctly? this works, but feels wrong. the structure is something like:
```json
{
  "navbar": [{"id": "explore", "hideFromNavbar": true}]
}
```

ideally we'd define some structs, like:
```go
type NavbarPreference struct {
  Id: string `json:"id"`
  HideFromNavbar bool `json:"hideFromNavbar"`
}

type PreferencesJsonData struct {
  Navbar NavbarPreference[] `json:"navbar"`
}
```

and use those, but i can't quite figure out the syntax 🤡

Reply:
Then maybe it's more optimal not converting read data from database to simplejson.Json and instead convert to the structure you want/need.

But with your current approach you should be able to something like:
```go
	jsonDataBytes, err := jsonDataPref.MarshalJSON()
	if err != nil {
		return err
	}
	var v PreferencesJsonData
	if err := json.Unmarshal(jsonDataBytes, &v); err != nil {
		return err
	}
```

Reply:
For the potential more optimal approach, might be able to do something like this (but haven't tested it):
```go
type Preferences struct {
	Id              int64
	OrgId           int64
	UserId          int64
	TeamId          int64
	Version         int
	HomeDashboardId int64
	Timezone        string
	WeekStart       string
	Theme           string
	Created         time.Time
	Updated         time.Time
	JsonData        *PreferencesJsonData
}

// The following needed for to implement the xorm/database ORM Conversion interface do the conversion when reading/writing to the database.

func (j *PreferencesJsonData) FromDB(data []byte) error {
	dec := json.NewDecoder(bytes.NewBuffer(data))
	dec.UseNumber()
	return dec.Decode(j)
}

func (j *PreferencesJsonData) ToDB() ([]byte, error) {
	if j == nil {
		return nil, nil
	}

	return json.Marshal(j)
}
```